### PR TITLE
Add data processing functionality to convert between data formats

### PR DESCRIPTION
**Pull request by RNKuhns**
This is related to #532 

#### Reference Issues/PRs

#532

#### What does this implement/fix? Explain your changes.

I've tried to round out the data container functionality to go between Pandas multi-indexed DataFrames, NumPy 3d arrays and Sktime nested DataFrame. I've also tried to have the conversions from nested to other structures (multi-indexed of NumPy 3d arrays) handle mixed DataFrames with some nested and some primitive columns. 

Also started on function nested_dataframes_equal for checking if two nested DataFrames (can be mix of nested an primitive columns) are equal.

Added functions:
from_multi_index_to_3d_numpy
from_3d_numpy_to_multi_index
from_multi_index_to_nested
from_nested_to_multi_index
_is_nested_structure
_make_column_names
nested_dataframes_equal

Updated functions:
_get_column_names to use _make_column_names.
is_nested_dataframe to account for mix of primitive
and nested column structures.
from_nested_to_3d_numpy to work with mix of
primitive and nested column structures and
handle nested structures
with  np.ndarrary or pd.Series.

Other changes are to formatting or docs.

#### What should a reviewer concentrate their feedback on?

This is my first time contributing to a project, so in addition to feedback on the functionality I've added, let me know anything that is out of sorts from a logistical standpoint in terms of contributing (I've tried to follow along with contrib guidelines).

I've also got a first draft of a function called nested_dataframes_equal to compare two potentially nested DataFrames for equality. Need to add check of columns and indexes (along with indexes of nested data structures), but interested in initial feedback. 

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/master/.all-contributorsrc).
- [ ] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/master/CODEOWNERS) to receive notifications about future changes to these files.
- [x] I've added unit tests and made sure they pass locally.

##### For new estimators
- [ ] I've added the estimator to the online documentation.
- [ ] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.

### CODE REVIEW

**Comment by RNKuhns** [comment]
@mloning This is my first time doing this, so I'm still learning how this works. Still need to add some unit tests and do a final review of documentation. But wanted to get your feedback.  This in regard to issue #532.

**Review by mloning — COMMENTED** [review]
Thanks @RNKuhns - that looks like a lot of useful code!

* Please check out the CONTRIBUTING.md section on code quality - let me know if you have any questions or get stuck, I'm happy to help!
* Can we add some simple unit tests? I'm sure you have written them already when developing these functions. Again, let me know if you need help with `pytest`. 
* We recently changed the file name from `data_container` to `data_processing` for the 0.5.0 release - that's why it tells you there is a merge conflict

**Review by KH633 — APPROVED** [review]

**Comment by RNKuhns** [comment]
I'm working to sync my fork with the v0.5 updates and also adding the pre-commit workflow and unit testing. 

I've run into an unexpected difficulty with the pre-commit workflow. it is stopping me from syncing my fork. It looks like everything is working except for 2 B301 exceptions from pre-commits flake8 checks. I've copied the screengrab below. Note that the use of `iter` in the method on those lines is just me using Pandas .iteritems() method on a DataFrame.

Note that I could do avoid the use of this by iterating over the DataFrames column names and slicing the values, but that doesn't seem like it it is the point of whats being checked. Let me know your thoughts on how to proceed. 

image

**Comment by RNKuhns** [comment]
I'm also running into an issue getting the development version of sktime installed so I can run the tests. 

I'm interpreting the CONTRIBUTING.md file to mean that I need to have the correct dependencies as specified in build_tools/requirements.txt. 

It isn't explictly mentioned after that, but I'm guessing that I then need to install the development version of sktime. I tried python setup.py install from the root directory, but I'm getting an error. 

The error I'm getting is "gcc: error: /openmp: No such file or directory. ". @mloning have you run into this with others in the past and do you have any advice for getting it corrected (note: I've checked and it looks like I've got intel-openmp installed). 

If it helps, I'm using Anaconda and have a virtual environment set up for the this that was built using build_tools/requirements.txt.

Thanks in advance for the help. Like I mentioned, I'm new to contributing to open-source projects, so I'm also learning how all the logistics work as I go.

**Review by Irapuato — APPROVED** [review]

**Comment by mloning** [comment]
Hi @RNKuhns, which operating system are you using? 

Having anaconda all set up is great. There was an error in the CONTRIBUTING.md. I've updated it now. See if you can work through [this section](https://github.com/alan-turing-institute/sktime/blob/master/CONTRIBUTING.md#continuous-integration) here and tell me where it fails.

**Review by KH633 — APPROVED** [review]

**Comment by RNKuhns** [comment]
@mloning I'm on Windows 10. 

I'm getting the error when I try to run "pip install -e .". of my local git repository copy of sktime. 

It is showing an error. I'm omitting the part that includes my system paths, etc. But the image below captures the main gist. 

image

**Comment by RNKuhns** [comment]
@mloning I also have a question about how to update this pull request. I pulled down the latest version of sktime and merged it into my fork. I've also added the unit tests we talked about. All of them are passing when the CI checks for Python 3.6 are run, but I'm having a failure for one with Python 3.7. I'm going to investigate and hopefully have that settled by the time this all gets straightened out.

**Comment by mloning** [comment]
@RNKuhns I'm assuming you have followed the advanced installation instructions [here](https://www.sktime.org/en/latest/installation.html). It may be worth having a look at scikit-learn's instructions [here](https://scikit-learn.org/stable/install.html), we're largely following their instructions. 

To update this PR, simply push again to the same branch.

**Comment by RNKuhns** [comment]
@mloning I'll push again to the same branch to update the PR. I'll also keep investigating what's up with that one test failing on Python 3.7 (even though it passes on Python 3.6). 

In terms of being able to install the development version so I can run the tests locally:

I've followed detailed instructions by specifying the path to thecompiler, which helps with the error I was seeing about  numpy.distutils._msvccompiler. However, I'm still getting what I hope is one last error (copied the text). It appears to be OpenMP related. I saw the info about setting the environment variable SKTIME_NO_OPENMP, but I am trying to avoid that.

I also didn't see anything about OpenMP on the Windows detailed installation instructions. Googling around has also yielded no answers so far. Let me know if you have any ideas on how to fix it. 

compiling C sources
C compiler: gcc -g -DDEBUG -DMS_WIN64 -O0 -Wall -Wstrict-prototypes

compile options: '-D__MSVCRT_VERSION__=0x1916 -IC:\MyAnacondaPath\anaconda3\envs\sktime_dev\lib\site-packages\numpy\core\include -IC:\MyAnacondaPath\anaconda3\envs\sktime_dev\lib\site-packages\numpy\core\include -IC:\MyAnacondaPath\anaconda3\envs\sktime_dev\include -IC:\MyAnacondaPath\anaconda3\envs\sktime_dev\include -c'
extra options: '/openmp'

gcc -g -DDEBUG -DMS_WIN64 -O0 -Wall -Wstrict-prototypes -D__MSVCRT_VERSION__=0x1916 -IC:\MyAnacondaPath\anaconda3\envs\sktime_dev\lib\site-packages\numpy\core\include -IC:\MyAnacondaPath\anaconda3\envs\sktime_dev\lib\site-packages\numpy\core\include -IC:\MyAnacondaPath\anaconda3\envs\sktime_dev\include -IC:\MyAnacondaPath\anaconda3\envs\sktime_dev\include -c sktime\distances\elastic_cython.c -o build\temp.win-amd64-3.7\Release\sktime\distances\elastic_cython.o /openmp
gcc: error: /openmp: No such file or directory
error: Command "gcc -g -DDEBUG -DMS_WIN64 -O0 -Wall -Wstrict-prototypes -D__MSVCRT_VERSION__=0x1916 -IC:\MyAnacondaPath\anaconda3\envs\sktime_dev\lib\site-packages\numpy\core\include -IC:\MyAnacondaPath\anaconda3\envs\sktime_dev\lib\site-packages\numpy\core\include -IC:\MyAnacondaPath\anaconda3\envs\sktime_dev\include -IC:\MyAnacondaPath\anaconda3\envs\sktime_dev\include -c sktime\distances\elastic_cython.c -o build\temp.win-amd64-3.7\Release\sktime\distances\elastic_cython.o /openmp" failed with exit status 1

**Comment by mloning** [comment]
@RNKuhns try this: 

1. Install anaconda
2. Create new environment: `conda create -n sktime-dev python=3.8`
3. Activate environment: `conda activate sktime-dev`
3. Install requirements: `pip install -r build_tools/requirements.txt`
4. Follow the instructions [here](https://scikit-learn.org/stable/developers/advanced_installation.html#windows) replacing `scikit-learn` with `sktime`

**Comment by RNKuhns** [comment]
Thanks for the help @mloning. I've had partial success!

I've went through the steps from your last post and was able to get the copy of sktime installed from my git repository. However, I ran into a issue -- fbprophet wouldn't build and install when attempting to install it from pip. I've left it off for now. It seems like it would be a soft dependency since it only affects certain parts of the module. 

The steps you laid out were very similar to what I did with a couple exceptions: 

1. I had conda install the packages if the conda versions were available (true for everything but catch22, pmdarima [windows version], and tbats)
2. I pip installed catch22, pmdarima and tbats
3. I then followed the rest of the instructions

I noticed that when I pip installed NumPy I initially got a message about NumPy 1.19.4 not working (see: https://developercommunity.visualstudio.com/content/problem/1207405/fmod-after-an-update-to-windows-2004-is-causing-a.html). 

I had to repeat the environment creation but first install NumPy pinned to 1.19.3. I had recently updated my Windows laptop after not having updated it for a while (I was wrapping up my semester and not wanting to mess anything up). 

Not sure if it is the case that the Windows update was causing a problem with NumPy and I had version 1.19.4 from Conda or if I was having a problem because I was using the Conda version of the packages. 

I'll try my initial approach with using the Conda packages (but pinning NumPy to 1.19.3) and let you know if that works or not in case anyone else runs into the same issue.

**Comment by mloning** [comment]
@RNKuhns thanks for the feedback, it'd be great if you could add this to the installation instructions with my list [here](https://github.com/alan-turing-institute/sktime/blob/master/docs/source/installation.rst)! 

`fbprophet` requires some extra work on Windows, check their installation instructions. 

I'm glad it's working now!

**Comment by RNKuhns** [comment]
@mloning I'll add those updates to the docs now. I see that 11 checks were successful and 1 is pending. Other than updating the commit to include the updated docs, is there anything else I need to do related to the pull request?

**Comment by mloning** [comment]
@RNKuhns great, looks all good for now, I'll try to review more thoroughly in the next few days!

File: sktime/utils/data_io.py
Comment (mloning) [inline sktime/utils/data_io.py:1034]:
```suggestion
def make_multi_index_dataframe
```

File: sktime/utils/data_processing.py
Comment (mloning) [inline sktime/utils/data_processing.py:730]: For this we already have a function [here](https://github.com/alan-turing-institute/sktime/blob/30b06140eb6e94b2845e6239efd679414dc13188/sktime/utils/_testing/estimator_checks.py#L641). Could you compare them and see if we can combine them into a single function if necessary?

**Review by mloning — COMMENTED** [review]
@RNKuhns this looks very good - thanks for all the work - this should make it quite a bit easier to go back and forth between data formats!

We need to add them to the [docs](https://github.com/alan-turing-institute/sktime/blob/master/docs/source/api_reference/utils.rst) too.

Adding a simple example notebook that shows some of these formats and how to convert them would be very helpful too I think, at least for the main formats: 3d numpy, nested pandas and the long format. What do you think?

**Comment by RNKuhns** [comment]
@mloning I'm down to add to the docs and put a sample notebook together. Might take me a week or so to finish b/c I'm working on a timeseries clustering/classification project concurrently.

**Comment by mloning** [comment]
Ideally, I'd like to have a single comparison function with options to select different checks (column type, index of series in cells, etc) but that may take some more work. Would you be interested in combining them? 

Extending the notebook would be very useful to users, we've been asked many times to provide better examples for data handling!

**Review by mloning — DISMISSED** [review]
Hi @RNKuhns - I think I'm happy to merge this as it is. Are you ready to merge or are you thinking about adding things?

**Comment by RNKuhns** [comment]
@mloning I got the updated examples finished and think that worked out pretty well. I also moved the from_long_to_nested from `sktime.utils.data_io` to `sktime.utils.data_processing`. 

I need to make 3 minor adjustments I just noticed when doing a final review,  but then I agree this is ready to go.  The adjustments I'm planning to make are:

1.  adding a unit test for  `from_long_to_nested` and `from_nested_to_long`.  
2. Update docs to include for `from_long_to_nested` now that I moved it to data_processing.  
3. I also think we remove the function `nested_dataframes_equal` for the time being. I can open an issue about modifying the existing function and do that after I have time to plan that out more. Is that fine with you? 

If that all sounds good to you, I'll stage my updates and run all the unit tests locally before pushing this to this pull request one last time. Should be able to send it by later today.

**Comment by mloning** [comment]
@RNKuhns yes, that sounds like a plan! Ping me when you're ready!

**Comment by RNKuhns** [comment]
@mloning I made the changes we discussed and updated the pull requests. The tests ran locally, so this will hopefully build and be ready to go in a bit.

**Review by mloning — COMMENTED** [review]
@RNKuhns could you take a look at the merge conflict with master. Once that's sorted, we can merge! 

FYI @TonyBagnall - since we talked about this yesterday, this implements a number of data conversion functions

**Comment by TonyBagnall** [comment]
this is great, I am prioritising sorting some of the I/O out and this will help a lot. I am happy for this to come in once conflicts are resolved and Markus is happy. I may do a bit of refactoring later, but this is really helpful

**Comment by RNKuhns** [comment]
@mloning I'll take a look today.

**Comment by RNKuhns** [comment]
@mloning I ran into a small issue resolving the merge conflicts (it was my first time doing that so I initially created an error), but think I got things taken care of.

**Comment by RNKuhns** [comment]
@mloning I initially ran into a bit of trouble with this (I managed to introduce a bug in my merge conflict resolution), but it should be all sorted now.

**Review by mloning — COMMENTED** [review]
Hi @RNKuhns, seems like you've accidentally added lots of auto-generated doc files (e.g. `docs/source/api_reference/modules/auto_generated/sktime.transformations.series.boxcox.BoxCoxTransformer.rst`). 

I think `docs/source/api_reference/modules/auto_generated/` should be added to the `.gitignore` file, an oversight on our part. Can you have to try to remove them again? Using the `git rm` command should be useful. Otherwise I can take a look!

**Review by mloning — APPROVED** [review]
Brilliant, thanks @RNKuhns looks all good to me now!

**Review by TonyBagnall — COMMENTED** [review]
looks good, thanks

**Pull request merged**
